src/main/java/baritone/process/elytra/NetherPathfinderContext.java

```diff
@@ -0,0 +1,239 @@
+import dev.babbaj.pathfinder.NetherPathfinder;
+import dev.babbaj.pathfinder.Octree;
+import dev.babbaj.pathfinder.PathSegment;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.init.Blocks;
+import net.minecraft.util.BitArray;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.ChunkPos;
+import net.minecraft.util.math.Vec3d;
+import net.minecraft.world.chunk.BlockStateContainer;
+import net.minecraft.world.chunk.Chunk;
+import net.minecraft.world.chunk.storage.ExtendedBlockStorage;
+
+import java.lang.ref.SoftReference;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * @author Brady
+ */
+public final class NetherPathfinderContext {
+
+    private static final IBlockState AIR_BLOCK_STATE = Blocks.AIR.getDefaultState();
+    public final Object cacheLock = new Object();
+
+    // Visible for access in BlockStateOctreeInterface
+    final long context;
+    private final long seed;
+    private final ExecutorService executor;
+
+    public NetherPathfinderContext(long seed) {
+        this.context = NetherPathfinder.newContext(seed);
+        this.seed = seed;
+        this.executor = Executors.newSingleThreadExecutor();
+    }
+
+    public void queueCacheCulling(int chunkX, int chunkZ, int maxDistanceBlocks, BlockStateOctreeInterface boi) {
+        this.executor.execute(() -> {
+            synchronized (this.cacheLock) {
+                boi.chunkPtr = 0L;
+                NetherPathfinder.cullFarChunks(this.context, chunkX, chunkZ, maxDistanceBlocks);
+            }
+        });
+    }
+
+    public void queueForPacking(final Chunk chunkIn) {
+        final SoftReference<Chunk> ref = new SoftReference<>(chunkIn);
+        this.executor.execute(() -> {
+            // TODO: Prioritize packing recent chunks and/or ones that the path goes through,
+            //       and prune the oldest chunks per chunkPackerQueueMaxSize
+            final Chunk chunk = ref.get();
+            if (chunk != null) {
+                long ptr = NetherPathfinder.getOrCreateChunk(this.context, chunk.x, chunk.z);
+                writeChunkData(chunk, ptr);
+            }
+        });
+    }
+
+    public void queueBlockUpdate(BlockChangeEvent event) {
+        this.executor.execute(() -> {
+            ChunkPos chunkPos = event.getChunkPos();
+            long ptr = NetherPathfinder.getChunkPointer(this.context, chunkPos.x, chunkPos.z);
+            if (ptr == 0) return; // this shouldn't ever happen
+            event.getBlocks().forEach(pair -> {
+                BlockPos pos = pair.first();
+                if (pos.getY() >= 128) return;
+                boolean isSolid = pair.second() != AIR_BLOCK_STATE;
+                Octree.setBlock(ptr, pos.getX() & 15, pos.getY(), pos.getZ() & 15, isSolid);
```

Comment:
Does this need to `synchronize` on `cacheLock` to ensure that `ptr` is still valid and not culled?

Reply:
not inside the executor